### ISSUE TITLE
Cluster Move

### ISSUE THREAD

**Issue by tuommaki** [giantswarm/roadmap#246]
# Cluster Move

## TL;DR;

Cluster Move feature provides functionality to move identity and workload from one cluster to another. It replaces upgrade and migration and helps with control plane reconfiguration.

### Cluster upgrades

Pretty often there is need to perform some kind of migration of resources, whether it's reconfiguration of load balancer or ASG / VMSS configuration. Maybe some form of refactoring of CloudFormation / ARM template structure. Generally a change that requires extra work to be programmed in such a way that customer workload doesn't get disrupted during the upgrade process.

If one could move cluster identity (i.e. DNS names and identifiers) PKI (customer access to cluster) would follow along. When being also able to bring existing workload to new cluster dynamically with basic kubernetes functionality, then there would be no need for upgrade or migration. Upgrade would be effectively a move to new cluster.

### Major infrastructure refactoring

Occasionally there are situations when major infrastructure refactoring is absolutely necessary. One time this happened was when node pools were implemented on AWS. Due to networking restrictions in AWS, we had to refactor
the way how subnets are managed for workload cluster and because of the fundamental nature of this it meant that there was effectively no upgrade path. This caused us a lot of issues on AWS because customers had to move to new clusters and start from scratch. There was no automatic migration available. Because of this we are still maintaining many old versions of AWS clusters even though node pool clusters have been released over 9 months ago already (as of writing).

### Control Plane move

Rarely there is situation where customer needs to move to new control plane. This happened with one customer when they hit the limits of existing CP network size and this happened with another customer when they split into own CP of their own.

While not happening so often, it would be convenient if we could move cluster from one control plane to another. Technically, as long as the provider is the same, there is not that much difference to plain cluster move within the same control plane. Major difference would be that there must be coordinated move of CP CRs from one CP to another, but after that normal reconciliation should theoretically fix the peerings etc.

#### Cluster ID / Name

Cluster ID / Name is the primary identifier across services both on control plane kubernetes but also on cloud provider. It is used for several different purposes and pretty much everything would break if that changed. There is no need to change it though so that's all right.

#### DNS

Similar to cluster identity, the DNS addresses for cluster must not change either. DNS would be one of the building blocks for making cluster move possible.

#### ETCD

Workload cluster `etcd` contains the whole cluster state so that must be preserved quite obviously.

#### Workload cluster PVCs

Disks provisioned for workload cluster workload must be preserved one way or another. More importantly, the data and availability of it must be preserved, not necessarily the specific instance of PVC disk.

#### 1. Create a new barebone control plane

Barebone control plane with new virtual cloud entity (VPC/Resource Group), network and similar setup.

#### 2. Create peering between old and new control plane networks

There must be network connectivity between old control plane and new one.

#### 3. Extend old control plane by creating additional master nodes to new one

Create new master nodes on new control plane so that they join the clustering of old control plane. Most importantly this would mean that etcd cluster is extended to cover both entities.

#### 4. Create matching node pools to new cluster entity

On new control plane / cluster entity create new node pools that mirror the ones in the old cluster. Ensure that they join the cluster flawlessly.

#### 5. Adjust the DNS entries so that they point to LBs of both cluster entities

In order to start receiving traffic to new cluster, DNS entry should be appended with LB of new cluster.

#### 6. Migrate the workload that has disks (PVCs)

Migrate the workload that has PVCs allocated for them. Precise method for this requires further investigation, but potentially scaling down the workload temporarily, taking a snapshot of disk, restoring it in the new cluster entity, patching PVC with this information and then scaling up the workload again in new node pools could be a solution.

#### 7. Migrate rest of the workload

Now rest of the workload should also be migrated to new node pool nodes.

#### 8. Adjust DNS entries so that they point only to LBs of new cluster

Once all the workload is running on new cluster, the DNS entry should also be adjusted accordingly.

#### 9. Scale down Etcd cluster to nodes running on new cluster entity

After scaling the Etcd cluster to span over both cluster entities, now that the move has been completed, the Etcd cluster must be scaled down so that it's fully running inside the new cluster entity.

#### 10. Drop the old cluster

Once the DNS TTL has expired and all traffic is flowing to new cluster, the old one can be disposed.

### Challenges

- Moving workload with disks is a challenge. There might be inevitable downtime due to this, but theoretically that should not be too long.

- Reliable etcd cluster move. Dynamically resizing etcd cluster should be doable, but requires extensive testing so that cluster state is not lost during move. There are always backups of course, but losing state would make things very complicated.

**Comment by cokiengchiara** [comment]
SIG Product Mon, 03. May 2021
* Replaces upgrades
* Will be talked about in the KaaS Sync - Team Celestial is working on a proposal

**Comment by puja108** [comment]
I'd like to revisit this once we're fully on CAPI. Furthermore, this is still a very big epic that might need quite a strategic push to follow. Might even be rather an RFC than already suited for considering in the roadmap in 6-12 month.

**Comment by cornelius-keller** [comment]
This feature will not be coming.
